src/TypeAutomata/FromAutomaton.hs

```diff
@@ -110,28 +122,40 @@ argNodesToArgTypes argNodes rep = do
     case ns of
       (Prd, ns) -> do
          typs <- forM ns (nodeToType rep)
+         let knd = checkTypKinds typs
          pure $ PrdCnsType PrdRep $ case rep of
-                                       PosRep -> mkUnion defaultLoc defaultKind typs
-                                       NegRep -> mkInter defaultLoc defaultKind typs
+                                       PosRep -> mkUnion defaultLoc knd typs
+                                       NegRep -> mkInter defaultLoc knd typs
       (Cns, ns) -> do
          typs <- forM ns (nodeToType (flipPolarityRep rep))
+         let knd = checkTypKinds typs
          pure $ PrdCnsType CnsRep $ case rep of
-                                       PosRep -> mkInter defaultLoc defaultKind typs
-                                       NegRep -> mkUnion defaultLoc defaultKind typs
+                                       PosRep -> mkInter defaultLoc knd typs
+                                       NegRep -> mkUnion defaultLoc knd typs
+
+checkTypKinds :: [Typ pol] -> MonoKind
```

Comment:
Throw monadic errors in this function, not error thunks.

Reply:
Changed it to monadic errors